File: Engine/ceCore/graphics/material/materialinstance.cc

#include <ceCore/graphics/material/materialinstance.hh>



namespace ce
{



MaterialInstance::MaterialInstance()
  : iMaterial()
  , m_material(nullptr)
{
  CE_CLASS_GEN_CONSTR;
}


MaterialInstance::~MaterialInstance()
{

}

bool MaterialInstance::Bind(iDevice* device, eRenderPass pass)
{
  if (!m_material)
  {
    return false;
  }
  if (!m_material->BindShader(device, pass))
  {
    return false;
  }

  m_material->BindBlending(device);
  m_material->BindDepthMode(device);
  m_material->BindFillMode(device);

  device->ResetTextures();
  for (Size i = 0, in = m_attributes.size(); i < in; i++)
  {
    Attribute& attribute = m_attributes[i];
    if (attribute.Override)
    {
      m_material->BindAttribute(device, pass, (uint16_t)i, attribute.Floats, attribute.Ints, attribute.Texture);
    }
    else
    {
      m_material->BindAttribute(device, pass, (uint16_t)i);
    }
  }
  device->MarkTexture();
  return true;
}

void MaterialInstance::SetMaterial(Material* material)
{
  CE_SET(m_material, material);
  RebuildAttributes();
}

Material* MaterialInstance::GetMaterial()
{
  return m_material;
}

const Material* MaterialInstance::GetMaterial() const
{
  return m_material;
}


Size MaterialInstance::IndexOf(const std::string& attributeName)
{
  return m_material ? m_material->IndexOf(attributeName) : Material::UndefinedIndex;
}



void MaterialInstance::Set(Size idx, float value)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  attr.Floats[0] = value;
}


void MaterialInstance::Set(Size idx, const Vector2f& v)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  attr.Floats[0] = v.x;
  attr.Floats[1] = v.y;
}


void MaterialInstance::Set(Size idx, const Vector3f& v)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  attr.Floats[0] = v.x;
  attr.Floats[1] = v.y;
  attr.Floats[2] = v.z;
}

void MaterialInstance::Set(Size idx, const Vector4f& v)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  attr.Floats[0] = v.x;
  attr.Floats[1] = v.y;
  attr.Floats[2] = v.z;
  attr.Floats[3] = v.w;
}


void MaterialInstance::Set(Size idx, const Color4f& v)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  attr.Floats[0] = v.r;
  attr.Floats[1] = v.g;
  attr.Floats[2] = v.b;
  attr.Floats[3] = v.a;
}

void MaterialInstance::Set(Size idx, int value)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  attr.Ints[0] = value;
}


void MaterialInstance::Set(Size idx, const Matrix3f& m)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  memcpy(attr.Floats.data(), &m, sizeof(float) * 9);
}

void MaterialInstance::Set(Size idx, const Matrix4f& m)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  memcpy(attr.Floats.data(), &m, sizeof(float) * 16);
}

void MaterialInstance::Set(Size idx, iTexture* texture)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  Attribute& attr = m_attributes[idx];
  attr.Override = true;
  CE_SET(attr.Texture, texture);

}

bool MaterialInstance::IsOverridden(Size idx) const
{
  if (idx >= m_attributes.size())
  {
    return false;
  }
  const Attribute& attr = m_attributes[idx];
  return attr.Override;
}

void MaterialInstance::SetOverride(Size idx, bool _override)
{
  if (idx >= m_attributes.size())
  {
    return;
  }
  m_attributes[idx].Override = _override;
}

void MaterialInstance::RebuildAttributes()
{
  m_attributes.clear();
  if (m_material)
  {
    std::vector<std::string> names = m_material->GetAttributeNames();
    for (auto name : names)
    {
      Attribute attribute;
      attribute.Override = false;
      attribute.Texture = nullptr;
      m_attributes.push_back(attribute);
    }
  }
}


const iShader *MaterialInstance::GetShader(eRenderPass pass) const
{
  return m_material ? m_material->GetShader(pass) : nullptr;
}

eFillMode MaterialInstance::GetFillMode() const
{
  return m_material ? m_material->GetFillMode() : eFillMode::Fill;
}

eRenderQueue MaterialInstance::GetRenderQueue() const
{
  return m_material ? m_material->GetRenderQueue() : eRenderQueue::Default;
}

eShadingMode MaterialInstance::GetShadingMode() const
{
  return m_material ? m_material->GetShadingMode() : eShadingMode::Shaded;
}

}

File: JCryoStorm/CsCore/src/main/java/org/cryo/core/graphics/material/MaterialInstance.java
package org.cryo.core.graphics.material;

import org.cryo.core.CsClass;
import org.cryo.core.CsObject;
import org.cryo.core.graphics.*;
import org.cryo.core.graphics.shading.IShader;
import org.cryo.core.math.Color4f;
import org.cryo.core.math.Vector2f;
import org.cryo.core.math.Vector3f;
import org.cryo.core.math.Vector4f;

import static org.cryo.core.graphics.material.MaterialInstanceNative.*;

@CsClass(CS_CLASS_NAME)
public class MaterialInstance extends IMaterial {

    public MaterialInstance() {
        super();
    }

    public MaterialInstance(long ref) {
        super(ref);
    }

    public IShader getShader(ERenderPass pass) {
        return (IShader) nGetShader(getRef(), pass.ordinal());
    }


    public EFillMode getFillMode() {
        return EFillMode.values()[nGetFillMode(getRef())];
    }

    public ERenderQueue getRenderQueue() {
        return ERenderQueue.values()[nGetRenderQueue(getRef())];
    }

    public EShadingMode getShadingMode() {
        return EShadingMode.values()[nGetShadingMode(getRef())];
    }

    public boolean bind(IDevice device, ERenderPass pass) {
        return nBind(getRef(), device.getRef(), pass.ordinal());
    }

    public void setMaterial (Material material) {
        nSetMaterial(getRef(), material.getRef());
    }

    public Material getMaterial () {
        return (Material)nGetMaterial(getRef());
    }

    public long indexOf (String attributeName) {
        return nIndexOf(getRef(), attributeName);
    }

    public void setFloat (long idx, float val) {
        nSetFloat(getRef(), idx, val);
    }

    public void setVector2f(long idx, Vector2f v) {
        nSetVector2f(getRef(), idx, v.x, v.y);
    }

    public void setVector3f(long idx, Vector3f v) {
        nSetVector3f(getRef(), idx, v.x, v.y, v.z);
    }

    public void setVector4f(long idx, Vector4f v) {
        nSetVector4f(getRef(), idx, v.x, v.y, v.z, v.w);
    }

    public void setColor4f(long idx, Color4f v) {
        nSetColor4f(getRef(), idx, v.r, v.g, v.b, v.a);
    }

    public void setTexture (long idx, ITexture texture) {
        nSetTexture(getRef(), idx, texture.getRef());
    }

    public boolean isOverridden (long idx) {
        return nIsOverridden(getRef(), idx);
    }

    public void setOverride (long idx, boolean override) {
        nSetOverride(getRef(), idx, override);
    }
}



